### PR TITLE
general: MainTitle as a pipe and test imports refactoring.

### PR DESCRIPTION
**Pull request by zannkukai**
* Refactoring MainTitleService as a pipe. It allows to use it more easely
  into components because we don't need to import the service and create
  a repetitive method in each component to get a document main title.
* This commit also renamed the Checkin component test file according to
  component name (previously named Checkout).
* Refactoring test 'spec.ts' files to limit imports and declarations.
  As main modules (AppModule and CirculationModules) already define
  many imports and declarations, use them into the 'spec.ts' file
  instead of a lot of single imports.
* Refactor name of some private attributes adding a leading '_'.

## Code review check list

- [ ] Commit message template compliance.
- [ ] Commit message without typos.
- [ ] File names.
- [ ] Functions names.
- [ ] Functions docstrings.
- [ ] Unnecessary commited files?
- [ ] Extracted translations?

### CODE REVIEW

**Comment by jma** [comment]
Can you use a spell checker for you commit messages?

File: projects/admin/src/app/circulation/checkin/checkin.component.spec.ts [hunk 1]
```diff
@@ -23,13 +23,14 @@ import { TranslateModule } from '@ngx-translate/core';
...
 import { UserService } from '../../service/user.service';
+import { SharedPipesModule } from '../../shared/shared-pipes.module';
 import { ItemComponent } from '../item/item.component';
 import { ItemsListComponent } from '../items-list/items-list.component';
 import { CardComponent } from '../patron/card/card.component';
 import { CheckinComponent } from './checkin.component';
 
 
-describe('CheckoutComponent', () => {
+describe('CheckinComponent', () => {
```
Comment (jma) [inline projects/admin/src/app/circulation/checkin/checkin.component.spec.ts:33]: This is not described in the commit message.

File: projects/admin/src/app/circulation/requested-items-list/requested-items-list.component.html [hunk 2]
```diff
@@ -29,21 +27,21 @@
...
   </thead>
   <tbody *ngIf="items && items.length">
-    <tr [ngClass]="{'table-success': item.loan.state !== 'PENDING'}"
-        *ngFor="let item of items">
-      <td><span *ngIf="getMainTitle(item.document.title) != null">
-        {{ getMainTitle(item.document.title) }}</span></td>
+    <tr [ngClass]="{'table-success': item.loan.state !== 'PENDING'}" *ngFor="let item of items">
+      <td>
+        <span *ngIf="(item.document.title | mainTitle) != null">
+          {{ item.document.title | mainTitle }}
```
Comment (jma) [inline projects/admin/src/app/circulation/requested-items-list/requested-items-list.component.html:33]: idem

File: projects/admin/src/app/pipe/main-title.pipe.ts [hunk 3]
```diff
@@ -0,0 +1,23 @@
+import { Pipe, PipeTransform } from '@angular/core';
```
Comment (jma) [inline projects/admin/src/app/pipe/main-title.pipe.ts:1]: copyright is missing

File: projects/admin/src/app/record/detail-view/acquisition-order-detail-view/order-lines/acquisition-order-lines.component.html [hunk 4]
```diff
@@ -47,8 +47,9 @@ <h4 class="mt-4" translate>Order lines</h4>
         <!-- DOCUMENT -->
         <ng-container *ngIf="orderLine.metadata.document.pid | getRecord: 'documents' | async as document">
         <div class="col-sm-3">
-          <a [routerLink]="['/records', 'documents', 'detail', document.metadata.pid]">
-            {{ getMainTitle(document.metadata.title) }}
+          <a [routerLink]="['/records', 'documents', 'detail', document.metadata.pid]"
+             *ngIf="(document.metadata.title | mainTitle) !== null">
```
Comment (jma) [inline projects/admin/src/app/record/detail-view/acquisition-order-detail-view/order-lines/acquisition-order-lines.component.html:51]: idem

File: projects/admin/src/app/record/detail-view/acquisition-order-line-detail-view/acquisition-order-line-detail-view.component.html [hunk 5]
```diff
@@ -53,8 +53,9 @@ <h1 class="mb-3">{{ record.metadata.description }}</h1>
               Title
             </dt>
             <dd class="col-sm-7 col-md-8 mb-0">
-              <a [routerLink]="['/records', 'documents', 'detail', document.metadata.pid]" *ngIf="getMainTitle(document.metadata.title) != null">
-                {{ getMainTitle(document.metadata.title) }}
+              <a [routerLink]="['/records', 'documents', 'detail', document.metadata.pid]"
+                 *ngIf="(document.metadata.title | mainTitle) != null">
```
Comment (jma) [inline projects/admin/src/app/record/detail-view/acquisition-order-line-detail-view/acquisition-order-line-detail-view.component.html:57]: idem

File: projects/admin/src/app/record/detail-view/document-detail-view/document-detail-view.component.html [hunk 6]
```diff
@@ -24,8 +24,8 @@
 
     <!-- TITLE -->
     <ng-container *ngIf="record.metadata.title">
-      <h1 *ngIf="getMainTitle(record.metadata.title) !== null" class="pl-0">
-        {{ getMainTitle(record.metadata.title) }}
+      <h1 *ngIf="(record.metadata.title | mainTitle) !== null" class="pl-0">
+        {{ record.metadata.title | mainTitle }}
```
Comment (jma) [inline projects/admin/src/app/record/detail-view/document-detail-view/document-detail-view.component.html:28]: idem

File: projects/public-search/src/app/search-bar/search-bar.component.ts [hunk 7]
```diff
@@ -31,10 +32,10 @@ export class SearchBarComponent implements OnInit {
   }
 
   constructor(
-    private translateService: TranslateService,
-    private _mainTitleService: MainTitleService
+    private _translateService: TranslateService,
```
Comment (jma) [inline projects/public-search/src/app/search-bar/search-bar.component.ts:35]: Not described in the commit message.

File: projects/admin/src/app/circulation/patron/history/history-item/history-item.component.html [hunk 8]
```diff
@@ -25,12 +25,12 @@
         <i [ngClass]="{ 'fa-caret-down': !isCollapsed, 'fa-caret-right': isCollapsed }" class="fa" aria-hidden="true"></i>
       </button>
       <a [routerLink]="['/records','documents','detail', document.metadata.pid]"
-         *ngIf="getMainTitle(document.metadata.title) != null">
+         *ngIf="(document.metadata.title | mainTitle) != null">
```
[the remark this answers is not in the document]
Comment (zannkukai) [reply projects/admin/src/app/circulation/patron/history/history-item/history-item.component.html:28]: done

File: projects/admin/src/app/record/detail-view/document-detail-view/document-detail-view.component.html (hunk 6 above)
Comment (zannkukai) [reply projects/admin/src/app/record/detail-view/document-detail-view/document-detail-view.component.html:28]: done

File: projects/admin/src/app/record/detail-view/acquisition-order-line-detail-view/acquisition-order-line-detail-view.component.html (hunk 5 above)
Comment (zannkukai) [reply projects/admin/src/app/record/detail-view/acquisition-order-line-detail-view/acquisition-order-line-detail-view.component.html:57]: done

File: projects/admin/src/app/record/detail-view/acquisition-order-detail-view/order-lines/acquisition-order-lines.component.html (hunk 4 above)
Comment (zannkukai) [reply projects/admin/src/app/record/detail-view/acquisition-order-detail-view/order-lines/acquisition-order-lines.component.html:51]: done

File: projects/admin/src/app/circulation/checkin/checkin.component.spec.ts (hunk 1 above)
Comment (zannkukai) [reply projects/admin/src/app/circulation/checkin/checkin.component.spec.ts:33]: done

File: projects/admin/src/app/circulation/patron/pickup/pickup-item/pickup-item.component.html [hunk 9]
```diff
@@ -21,8 +21,8 @@
   </div>
   <!-- TITLE -->
   <div class="col-sm-5"><a [routerLink]="['/records', 'documents', 'detail', item.document.pid]"
-    *ngIf="getMainTitle(item.document.title) != null">
-    {{ getMainTitle(item.document.title) }}</a>
+    *ngIf="item.document.title | mainTitle as title">
+    {{ title | mainTitle }}</a>
```
Comment (jma) [inline projects/admin/src/app/circulation/patron/pickup/pickup-item/pickup-item.component.html:25]: you can remove the pipe here.

File: projects/admin/src/app/record/detail-view/person-detail-view/person-detail-view.component.html [hunk 10]
```diff
@@ -142,8 +142,8 @@ <h1 class="mb-0">{{ record.metadata | mefTitle }}</h1>
     <article id="collapse-rero" class="card-body collapse show">
       <ul class="list-unstyled mb-0" *ngFor="let document of documents">
         <li>
-          <a [routerLink]="['/records', 'documents', 'detail', document.metadata.pid]" *ngIf="getMainTitle(document.metadata.title) != null">
-            {{ getMainTitle(document.metadata.title) }}
+          <a [routerLink]="['/records', 'documents', 'detail', document.metadata.pid]" *ngIf="(document.metadata.title | mainTitle) != null">
```
Comment (jma) [inline projects/admin/src/app/record/detail-view/person-detail-view/person-detail-view.component.html:145]: you can use `as` here

File: projects/admin/src/app/circulation/item/item.component.ts [hunk 11]
```diff
@@ -66,8 +65,7 @@ export class ItemComponent implements OnInit {
   constructor(
     private recordService: RecordService,
     private organisationService: OrganisationService,
-    private patronTransactionService: PatronTransactionService,
-    private _mainTitleService: MainTitleService
+    private patronTransactionService: PatronTransactionService
```
[the remark this answers is not in the document]
Comment (zannkukai) [reply projects/admin/src/app/circulation/item/item.component.ts:68]: adding "_" for each constructor arguments.

File: projects/admin/src/app/circulation/main-request/main-request.component.spec.ts [hunk 12]
```diff
@@ -19,18 +19,22 @@ import { HttpClientModule } from '@angular/common/http';
...
   let component: MainRequestComponent;
   let fixture: ComponentFixture<MainRequestComponent>;
 
   beforeEach(async(() => {
     TestBed.configureTestingModule({
-      imports: [RecordModule, RouterTestingModule, HttpClientModule, TranslateModule.forRoot()],
-      declarations: [ MainRequestComponent, RequestedItemsListComponent ]
+      imports: [
+        RouterTestingModule,
+        HttpClientModule,
```
[the remark this answers is not in the document]
Comment (zannkukai) [reply projects/admin/src/app/circulation/main-request/main-request.component.spec.ts:34]: RouterTestingModule isn't required, HttpClientModule is

File: projects/admin/src/app/circulation/patron/card/card.component.spec.ts [hunk 13]
```diff
@@ -19,17 +19,22 @@ import { HttpClientModule } from '@angular/common/http';
...
   let fixture: ComponentFixture<CardComponent>;
 
   beforeEach(async(() => {
     TestBed.configureTestingModule({
-      imports: [TranslateModule, RouterTestingModule, HttpClientModule],
-      declarations: [ CardComponent ]
+      imports: [
+        TranslateModule.forRoot(),
+        RouterTestingModule,
+        HttpClientModule,
```
[the remark this answers is not in the document]
Comment (zannkukai) [reply projects/admin/src/app/circulation/patron/card/card.component.spec.ts:35]: Removed excepted HttpClientModule

File: projects/admin/src/app/circulation/patron/loan/loan.component.spec.ts [hunk 14]
```diff
@@ -19,40 +19,45 @@ import { HttpClientModule } from '@angular/common/http';
...
 
 
 describe('LoanComponent', () => {
   let component: LoanComponent;
   let fixture: ComponentFixture<LoanComponent>;
 
+  const userService = jasmine.createSpyObj(
+    'UserService', ['getCurrentUser']
+  );
+  userService.getCurrentUser.and.returnValue({
```
[the remark this answers is not in the document]
Comment (zannkukai) [reply projects/admin/src/app/circulation/patron/loan/loan.component.spec.ts:36]: removed

File: projects/admin/src/app/record/detail-view/item-detail-view/item-detail-view.component.ts [hunk 15]
```diff
@@ -53,41 +52,23 @@ export class ItemDetailViewComponent implements DetailRecord, OnInit {
...
-      ).subscribe();
-  }
-
-  /**
-   * Get main title (correspondig to 'bf_Title' type, present only once in metadata)
-   * @param titleMetadata: title metadata
-   */
-  getMainTitle(titleMetadata: any): string {
-    return this._mainTitleService.getMainTitle(titleMetadata);
+    this.record$.subscribe( record => {
```
[the remark this answers is not in the document]
Comment (zannkukai) [reply projects/admin/src/app/record/detail-view/item-detail-view/item-detail-view.component.ts:59]:
```
  ngOnDestroy(): void {
    this.record$.unsubscribe();
  }
````
cause 
` error TS2551: Property 'unsubscribe' does not exist on type 'Observable<any>'. Did you mean 'subscribe'?`

File: projects/admin/src/app/record/detail-view/document-detail-view/document-detail-view.component.ts [hunk 16]
```diff
@@ -15,22 +15,24 @@
...
   styles: []
 })
-export class DocumentDetailViewComponent implements DetailRecord, OnInit {
+export class DocumentDetailViewComponent implements DetailRecord, OnInit, OnDestroy {
 
   /** Observable resolving record data */
   record$: Observable<any>;
 
+  /** Record subscription */
+  recordSubscription: Subscription;
```
Comment (jma) [inline projects/admin/src/app/record/detail-view/document-detail-view/document-detail-view.component.ts:34]: you can mark this as private (do not forget the _)

File: projects/admin/src/app/record/detail-view/item-detail-view/item-detail-view.component.ts [hunk 17]
```diff
@@ -14,23 +14,24 @@
...
   templateUrl: './item-detail-view.component.html',
   styles: []
 })
-export class ItemDetailViewComponent implements DetailRecord, OnInit {
+export class ItemDetailViewComponent implements DetailRecord, OnInit, OnDestroy {
   /** Observable resolving record data */
   record$: Observable<any>;
 
+  /** Record subscription */
+  recordSubscription: Subscription;
```
Comment (jma) [inline projects/admin/src/app/record/detail-view/item-detail-view/item-detail-view.component.ts:33]: private?

File: projects/admin/src/app/record/detail-view/library-detail-view/library-detail-view.component.ts [hunk 18]
```diff
@@ -14,22 +14,25 @@
...
   styles: []
 })
-export class LibraryDetailViewComponent implements DetailRecord, OnInit {
+export class LibraryDetailViewComponent implements DetailRecord, OnInit, OnDestroy {
 
   /** Observable resolving record data */
   record$: Observable<any>;
 
+  /** Record subscription */
+  recordSubscription: Subscription;
```
Comment (jma) [inline projects/admin/src/app/record/detail-view/library-detail-view/library-detail-view.component.ts:34]: private?

File: projects/admin/src/app/record/detail-view/location-detail-view/location-detail-view.component.ts [hunk 19]
```diff
@@ -34,7 +35,7 @@ export class LocationDetailViewComponent implements OnInit {
   record: any;
 
   /** The observer to the record observable */
-  private recordObs = null;
+  private recordObs: Subscription;
```
Comment (jma) [inline projects/admin/src/app/record/detail-view/location-detail-view/location-detail-view.component.ts:38]: private check the others

File: projects/admin/src/app/record/detail-view/document-detail-view/document-detail-view.component.ts (hunk 16 above)
Comment (zannkukai) [reply projects/admin/src/app/record/detail-view/document-detail-view/document-detail-view.component.ts:34]: redeclared as `private _recordObs` in all component.

File: projects/admin/src/app/record/detail-view/item-detail-view/item-detail-view.component.ts (hunk 17 above)
Comment (zannkukai) [reply projects/admin/src/app/record/detail-view/item-detail-view/item-detail-view.component.ts:33]: same as previous

**Pull request merged**